### PR TITLE
Add `traverse` kwarg to `delayed`

### LINKED ISSUE #1884: calling delayed functions with lists is slow
The following is relatively slow:

```
myfun = lambda *args, **kwargs: None
%time d = [dask.delayed(myfun)(list(range(10000))) for i in range(100)]
```

> CPU times: user 11.2 s, sys: 88 ms, total: 11.3 s
> Wall time: 11.3 s

Since generating 100 lists of length 10k only takes about 50ms, the majority of the CPU time seems to be due to dask.delayed. What is dask.delayed doing behind the scenes?

Edit:
```
dask.__version__
> u'0.12.0'
```

### PR DESCRIPTION
**Pull request by jcrist**
By default dask traverses builtin python collections looking for dask
objects passed to ``delayed``. For large collections this can be
expensive. This PR adds a keyword ``traverse`` to ``delayed``, to tell
dask to avoid doing this traversal.

Fixes #1884.

### CODE REVIEW

File: dask/delayed.py [hunk 1]
```diff
@@ -248,9 +254,13 @@ def delayed(obj, name=None, pure=False, nout=None):
     if isinstance(obj, Delayed):
         return obj
 
-    task, dasks = to_task_dasks(obj)
+    if isinstance(obj, base.Base) or traverse:
+        task, dasks = to_task_dasks(obj)
+    else:
+        task = quote(obj)
+        dasks = []
```
Comment (mrocklin) [inline dask/delayed.py]: I'm curious, why the need to quote?

Comment (jcrist) [inline dask/delayed.py]: If `obj` is a tuple that looks like a task, or a list containing tuples that look like tasks. Looks like the second isn't handled though, will fix.

**Comment by jcrist** [comment]
Should be good to go now.

**Comment by jcrist** [comment]
Merging soon if no comment.

**Pull request merged**